packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,476 @@
+> = MethodImplementation extends (args: infer Options) => Json | Promise<Json>
+  ? Options extends RestrictedMethodOptions<RestrictedMethodParameters>
+    ? MethodImplementation
+    : never
+  : never;
+
+export type PermissionFactory<
+  TargetPermission extends GenericPermission,
+  RequestData extends Record<string, unknown>,
+> = (
+  options: PermissionOptions<TargetPermission>,
+  requestData?: RequestData,
+) => TargetPermission;
+
+export type PermissionValidator<TargetPermission extends GenericPermission> = (
+  permission: GenericPermission,
+  origin?: OriginString,
+  target?: TargetPermission['parentCapability'],
+) => void;
+
+/**
+ * A utility type for ensuring that the given permission target key conforms to
+ * our naming conventions.
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key string to apply the constraint to.
+ */
+type ValidateTargetKey<Key extends string> = Key extends `${string}_*`
+  ? Key
+  : Key extends `${string}_`
+  ? never
+  : Key extends `${string}*`
+  ? never
+  : Key;
+
+/**
+ * A type extended by any function.
+ */
+type GenericFunction = (...args: any[]) => any;
+
+export type PermissionSpecificationConstraint = {
+  /**
+   * The target resource of the permission. In other words, at the time of
+   * writing, the RPC method name.
+   */
+  targetKey: string;
+
+  /**
+   * An array of the caveat types that may be added to instances of this
+   * permission.
+   */
+  allowedCaveats?: Readonly<NonEmptyArray<string>>;
+
+  /**
+   * The implementation of the restricted method that the permission
+   * corresponds to.
+   */
+  methodImplementation: GenericFunction;
+
+  /**
+   * The factory function used to get permission objects. Permissions returned
+   * by this function are presumed to valid, and they will not be passed to the
+   * validator function associated with this specification (if any). In other
+   * words, the factory function should validate the permissions it creates.
+   *
+   * If no factory is specified, the {@link Permission} constructor will be
+   * used, and the validator function (if specified) will be called on newly
+   * constructed permissions.
+   */
+  factory?: PermissionFactory<any, Record<string, unknown>>;
+
+  /**
+   * The validator function used to validate permissions of the associated type
+   * whenever they are mutated. The only way a permission can be legally mutated
+   * is when its caveats are modified by the permission controller.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   */
+  validator?: PermissionValidator<GenericPermission>;
+};
+
+/**
+ * Constraint for {@link PermissionSpecificationConstraint} objects that
+ * evaluates to `never` if the specification contains any invalid fields.
+ *
+ * @template Specification - The permission specification to validate.
+ */
+export type ValidatePermissionSpecification<Specification> =
```

Comment:
Nit: we may as well move this first conditional into the generic parameter spot:

```suggestion
export type ValidatePermissionSpecification<Specification extends PermissionSpecificationConstraint> =
```

Should be the same except more concise, I think?